fvm/fvm.go

```diff
@@ -76,17 +95,50 @@ func (vm *VirtualMachine) RunV2(
 ) error {
 	blockPrograms := ctx.BlockPrograms
 	if blockPrograms == nil {
-		blockPrograms = programs.NewEmptyPrograms()
+		blockPrograms = programs.NewEmptyBlockProgramsWithTransactionOffset(
+			uint32(proc.ExecutionTime()))
+	}
+
+	var txnPrograms *programs.TransactionPrograms
+	var err error
+	if proc.Type() == ScriptProcedureType {
+		txnPrograms, err = blockPrograms.NewSnapshotReadTransactionPrograms(
+			proc.InitialSnapshotTime(),
+			proc.ExecutionTime())
+	} else {
+		txnPrograms, err = blockPrograms.NewTransactionPrograms(
+			proc.InitialSnapshotTime(),
+			proc.ExecutionTime())
+	}
+
+	if err != nil {
+		return fmt.Errorf("error creating transaction programs: %w", err)
+	}
+
+	// Note: it is safe to skip committing the parsed programs for non-normal
+	// transactions (i.e., bootstrap and script) since this is only an
+	// optimization.
+	if proc.Type() == TransactionProcedureType {
+		defer func() {
+			commitErr := txnPrograms.Commit()
+			if commitErr != nil {
+				// NOTE: txnPrograms commit error does not impact correctness,
+				// but may slow down execution since some programs may need to
+				// be re-parsed.
+				ctx.Logger.Err(commitErr).Msg(
```

Comment:
Do we know what could possibly cause the failure? If the failure is normal and doesn't matter, maybe we could log as warn?
```suggestion
				ctx.Logger.Warn(commitErr).Msg(
```